README.md
# Eliza-GPT

[![Build status](https://github.com/miguelgrinberg/eliza-gpt/workflows/build/badge.svg)](https://github.com/miguelgrinberg/eliza-gpt/actions) [![codecov](https://codecov.io/gh/miguelgrinberg/eliza-gpt/branch/main/graph/badge.svg)](https://codecov.io/gh/miguelgrinberg/eliza-gpt)

[Eliza](https://en.wikipedia.org/wiki/ELIZA), the classic chatbot from the 1960s, running on OpenAI's [Chat Completions API](https://platform.openai.com/docs/api-reference/chat).

## Why?

The main reason is that this seemed like a fun project to help me acquire a greater understanding of the Chat Completions API from OpenAI, which has become a sort of standard for turn-based conversation between a human and a machine.

But also, it is great to have a fast and free service with negligible resource consumption that can be used as a stand-in for a real LLM when the quality of the responses do not matter, such as while doing development or testing.

## How To Use

Eliza-GPT is written in Python and can be installed with `pip`:

    $ pip install eliza-gpt

Once installed, run it to start it as a local service on your computer:


    $ eliza-gpt
    Eliza-GPT is running!
    Set base_url="http://127.0.0.1:5005/v1" in your OpenAI client to connect.

### Configuration

Run with `--help` to learn about configuration options, which include:

- Setting the listening IP address and port.
- Adding an API key for authentication.
- Changing the (simulated) response times. Set to 0 to have near immediate responses.

### Connecting With a Chat Client

If you are using the official Python client from OpenAI, add the `base_url` option to it as follows:

    openai_client = OpenAI(base_url='http://127.0.0.1:5005/v1')

If you use Langchain, then use the following to connect to Eliza-GPT:

    chat = ChatOpenAI(base_url='http://127.0.0.1:5005/v1')

If you have a custom client that talks directly to the chat completions endpoint, configure `http://127.0.0.1:5005/v1` as your endpoint.

Eliza-GPT supports both the direct and streaming interfaces.

## Examples

The [examples](https://github.com/miguelgrinberg/Eliza-GPT/tree/main/examples) directory includes demonstration applications implemented with the OpenAI client and Langchain.

## Implementation Details

Eliza-GPT implements a portion of the OpenAI Chat Completions API, ignoring anything that isn't useful. In particular, at this time only the `/v1/chat/completions` endpoint is implemented. Any other endpoint from the OpenAI API will return a 404 error to the caller.

The following aspects of the Chat Completions endpoint are currently supported:

- The chat history given in the `messages` argument. Only messages with role `user` are used to "prime" Eliza so that it provides a reasonable response.
- The `stream` argument, which determines if the response should be given in a single JSON object or as a stream of Server-Sent Events.
- The `seed` option, which makes it possible to have deterministic responses.

These are not used or not implemented:

- Any messages in the chat history that have a role different than `user`.
- Model names. Eliza-GPT returns the requested model name in the response.
- The `n` argument, which controls how many responses are returned. Eliza-GPT always returns a single response.
- The `max_tokens` argument.
- Temperature and other LLM-specific tuning parameters.
- The `stop` argument.
- Anything related to tools or functions that the model may call.
- Anything related to log probabilities.

## Credits

The Eliza chatbot used in this project is called [Eliza.py](https://github.com/rdimaio/eliza-py) and was created by Riccardo Di Maio.


CHANGES.md
# Eliza-GPT Change Log

**Release 0.2.0** - 2023-12-28

- Use microdot's SSE extension ([commit](https://github.com/miguelgrinberg/eliza-gpt/commit/0a1f3e087e97284c0c4dcd0ea766c7fdd5dd6a63))
- Documentation for the example applications ([commit](https://github.com/miguelgrinberg/eliza-gpt/commit/a559d14c2cf4bfbc60780186ccb004afd77a799b))

**Release 0.1.1** - 2023-12-24

- Fix issues with Python packaging ([commit](https://github.com/miguelgrinberg/eliza-gpt/commit/b6d885fdefe644d6863f2cc84a2b769a46e040b7))

**Release 0.1.0** - 2023-12-24

- Initial release


tests/test_eliza_gpt.py
import json
import pytest
from eliza_gpt import ElizaGPT
from eliza_gpt.cli import main  # noqa: F401
from microdot.test_client import TestClient


@pytest.mark.asyncio
async def test_eliza_gpt():
    app = ElizaGPT(api_key='test', avg_response_time=0.01)
    client = TestClient(app)

    r = await client.post('/v1/chat/completions')
    assert r.status_code == 401

    r = await client.post(
        '/v1/chat/completions', headers={'Authorization': 'Bearer test'},
    )
    assert r.status_code == 400

    r = await client.post(
        '/v1/chat/completions', headers={'Authorization': 'Bearer test'},
        body={
            'messages': [
                {'role': 'system', 'content': 'You are a helpful assistant.'},
            ],
        }
    )
    assert r.status_code == 400

    r = await client.post(
        '/v1/chat/completions', headers={'Authorization': 'Bearer test'},
        body={
            'messages': [
                {'role': 'system', 'content': 'You are a helpful assistant.'},
                {'role': 'user', 'content': 'I like Python.'},
            ],
        }
    )
    assert r.status_code == 200
    assert r.headers['Content-Type'].startswith('application/json')
    assert r.json['choices'][0]['message']['role'] == 'assistant'
    assert len(r.json['choices'][0]['message']['content']) > 0

    r = await client.post(
        '/v1/chat/completions', headers={'Authorization': 'Bearer test'},
        body={
            'messages': [
                {'role': 'system', 'content': 'You are a helpful assistant.'},
                {'role': 'user', 'content': 'I like Python.'},
            ],
            'stream': True,
            'seed': 123,
        }
    )
    assert r.status_code == 200
    assert r.headers['Content-Type'].startswith('text/event-stream')
    lines = r.body.splitlines()
    count = 0
    for line in lines:
        if line.startswith(b'data: '):
            if line == b'data: [DONE]':
                break
            count += 1
            chunk = json.loads(line[6:])
            assert chunk['choices'][0]['delta']['role'] == 'assistant'
            assert len(chunk['choices'][0]['delta']['content']) > 0
    assert count > 0


src/eliza_gpt/cli.py
import argparse
from eliza_gpt.api import ElizaGPT


def main():  # pragma: no cover
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--host', type=str, default='127.0.0.1',
        help='The host or IP address to listen on (default "127.0.0.1")')
    parser.add_argument(
        '--port', type=int, default=5005,
        help='The port to listen on (default 5005)')
    parser.add_argument(
        '--debug', action='store_true',
        help='Show incoming requests in the terminal')
    parser.add_argument(
        '--api-key', type=str, default=None,
        help='The API key to use for authentication')
    parser.add_argument(
        '--avg-response-time', type=float, default=2,
        help='The average response time in seconds (default 2 seconds)')
    args = parser.parse_args()

    print('Eliza-GPT is running!')
    print(f'Set base_url="http://{args.host}:{args.port}/v1" in your OpenAI '
          'client to connect.')
    app = ElizaGPT(api_key=args.api_key,
                   avg_response_time=args.avg_response_time)
    app.run(host=args.host, port=args.port, debug=args.debug)


examples/openai_chat.py
import os
from openai import OpenAI

openai_client = OpenAI(
    base_url='http://localhost:5005/v1',
    api_key=os.environ.get('OPENAI_API_KEY', 'x'),
)
messages = [
    {"role": "system", "content": "You are a helpful assistant."},
]
stream = True  # set to False to get all responses at once

while True:
    user_message = input("You: ")
    messages.append({"role": "user", "content": user_message})

    if stream:
        # this returns the responses in chunks
        stream = openai_client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=messages,
            stream=True,
        )
        print('Eliza: ', end='', flush=True)
        response = ''
        for chunk in stream:
            if chunk.choices[0].delta.content is not None:
                print(chunk.choices[0].delta.content, end='', flush=True)
                response += chunk.choices[0].delta.content
        print('')

    else:
        # this returns the response all at once
        completion = openai_client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=messages,
        )
        response = completion.choices[0].message.content
        print('Eliza:', response)

    messages.append({"role": "assistant", "content": response})


examples/langchain_chat.py
import os
from langchain.chat_models import ChatOpenAI
from langchain.schema import HumanMessage, SystemMessage, AIMessage

chat = ChatOpenAI(
    model="gpt-3.5-turbo",
    base_url='http://localhost:5005/v1',
    api_key=os.environ.get('OPENAI_API_KEY', 'x'),
)
messages = [
    SystemMessage(content="You are a helpful assistant."),
]
stream = True

while True:
    user_message = input("You: ")
    messages.append(HumanMessage(content=user_message))

    if stream:
        # get responses in chunks
        print('Eliza: ', end='', flush=True)
        response = ''
        for chunk in chat.stream(messages):
            if chunk.content is not None:
                print(chunk.content, end='', flush=True)
                response += chunk.content
        print('')
        response = AIMessage(content=response)
    else:
        # get responses all at once
        response = chat(messages)
        print('Eliza:', response.content)

    messages.append(response)


src/eliza_gpt/v1.py
import asyncio
import random
import time
import uuid
from microdot import Microdot
from microdot.sse import sse_response
from eliza_py.utils.startup import setup as eliza_setup
from eliza_py.utils.response import generate_response as run_eliza
from eliza_py.eliza import GENERAL_SCRIPT_PATH, SCRIPT_PATH

api = Microdot()
general_script, script, memory_inputs, _ = eliza_setup(
    GENERAL_SCRIPT_PATH, SCRIPT_PATH)
system_fingerprint = uuid.uuid4().hex


async def get_response_from_eliza(messages):
    memory_stack = []
    response = ''
    for msg in messages:
        response = run_eliza(msg, script, general_script['substitutions'],
                             memory_stack, memory_inputs)
        await asyncio.sleep(0)
    response = response.replace('Eliza: ', '')
    response = response.replace('\nYou: ', '')
    return response


@api.post('/chat/completions')
async def chat_completions(request):
    if request.app.api_key is not None:  # pragma: no branch
        if request.headers.get('Authorization') != \
                f'Bearer {request.app.api_key}':
            return {'error': 'Invalid API key'}, 401
    if request.json is None:
        return {'error': 'No data provided'}, 400

    messages = [msg['content'] for msg in request.json.get('messages', [])
                if msg['role'] == 'user']
    if len(messages) == 0:
        return {'error': 'No messages provided'}, 400

    chat_id = uuid.uuid4().hex
    created = int(time.time())
    model = request.json.get('model', 'eliza-gpt')
    seed = request.json.get('seed')
    if seed:
        random.seed(seed)
    response = await get_response_from_eliza(messages)

    if not request.json.get('stream', False):
        if request.app.avg_response_time:  # pragma: no branch
            await asyncio.sleep(
                random.random() * request.app.avg_response_time * 2)
        return {
            'id': chat_id,
            'created': created,
            'model': model,
            'system_fingerprint': system_fingerprint,
            'choices': [
                {
                    'index': 0,
                    'message': {
                        'role': 'assistant',
                        'content': response,
                    },
                    'finish_reason': 'stop',
                    'logprobs': None,
                }
            ]
        }

    async def send_chunks(request, sse):
        chunks = response.split(' ')
        for i in range(len(chunks)):
            last_chunk = i == len(chunks) - 1
            if request.app.avg_response_time:  # pragma: no branch
                await asyncio.sleep(
                    random.random() * request.app.avg_response_time * 2 / 10)
            await sse.send({
                'id': chat_id,
                'created': created,
                'model': model,
                'system_fingerprint': system_fingerprint,
                'object': 'chat.completion.chunk',
                'choices': [
                    {
                        'index': 0,
                        'delta': {
                            'role': 'assistant',
                            'content': chunks[i] + ('' if last_chunk else ' ')
                        },
                        'finish_reason': 'stop' if last_chunk else None,
                    },
                ],
            })
        await sse.send('[DONE]')

    return sse_response(request, send_chunks)
